File: scripts/update_baseline.py
#!/usr/bin/env python3
"""
更新性能基线脚本
用于更新性能基线文件
"""

import os
import sys
import json
import argparse
from datetime import datetime
import shutil


def load_current_results(results_file):
    """加载当前性能测试结果"""
    try:
        with open(results_file, 'r') as f:
            return json.load(f)
    except Exception as e:
        print(f"Error loading results: {e}", file=sys.stderr)
        return None


def load_baseline(baseline_file):
    """加载当前基线"""
    try:
        with open(baseline_file, 'r') as f:
            return json.load(f)
    except Exception as e:
        print(f"Baseline file not found, will create new: {e}", file=sys.stderr)
        return None


def extract_baseline_data(results, version=None, commit=None):
    """从结果中提取基线数据"""
    if not results:
        return None
    
    baseline_data = {
        'version': version or 'unknown',
        'date': datetime.utcnow().isoformat() + 'Z',
        'commit': commit or 'unknown',
        'baseline': {}
    }
    
    # 提取 RPS 数据
    if 'test_scenarios' in results:
        for scenario in results['test_scenarios']:
            scenario_name = scenario.get('name', 'unknown')
            
            if 'results' in scenario:
                scenario_data = {}
                
                if 'rps' in scenario['results']:
                    scenario_data['rps'] = scenario['results']['rps']['value']
                
                if 'latency_avg' in scenario['results']:
                    latency = scenario['results']['latency_avg']['value']
                    if latency != 'N/A':
                        scenario_data['latency_avg'] = latency
                
                if scenario_data:
                    baseline_data['baseline'][scenario_name] = scenario_data
    
    # 提取环境信息
    if 'environment' in results:
        baseline_data['environment'] = {
            'os': results['environment'].get('os', 'unknown'),
            'runner': results['environment'].get('runner', 'unknown'),
            'cpu': results['environment'].get('cpu', 'unknown'),
            'compiler': results['environment'].get('compiler', 'unknown'),
            'compiler_version': results['environment'].get('compiler_version', 'unknown'),
            'build_type': results['environment'].get('build_type', 'Release'),
            'optimization': results['environment'].get('optimization', '-O2')
        }
    
    return baseline_data


def update_baseline_file(baseline_file, new_baseline, backup=True):
    """更新基线文件"""
    # 备份旧基线
    if backup and os.path.exists(baseline_file):
        backup_file = f"{baseline_file}.backup"
        shutil.copy2(baseline_file, backup_file)
        print(f"Baseline backed up to {backup_file}")
    
    # 写入新基线
    with open(baseline_file, 'w') as f:
        json.dump(new_baseline, f, indent=2)
    
    print(f"Baseline updated: {baseline_file}")
    print(f"Version: {new_baseline['version']}")
    print(f"Date: {new_baseline['date']}")
    
    return True


def update_history_file(history_file, new_baseline, max_entries=30):
    """更新历史记录文件"""
    history = []
    
    # 加载现有历史
    if os.path.exists(history_file):
        try:
            with open(history_file, 'r') as f:
                history = json.load(f)
        except:
            history = []
    
    # 添加新条目
    history.insert(0, new_baseline)
    
    # 限制条目数量
    history = history[:max_entries]
    
    # 写入历史文件
    with open(history_file, 'w') as f:
        json.dump(history, f, indent=2)
    
    print(f"History updated: {history_file} ({len(history)} entries)")
    
    return True


def compare_with_baseline(new_baseline, old_baseline):
    """比较新基线与旧基线"""
    if not old_baseline:
        print("No old baseline found for comparison")
        return
    
    print("\n=== Baseline Comparison ===\n")
    
    new_benchmarks = new_baseline.get('baseline', {})
    old_benchmarks = old_baseline.get('baseline', {})
    
    all_scenarios = set(new_benchmarks.keys()) | set(old_benchmarks.keys())
    
    for scenario in sorted(all_scenarios):
        new_data = new_benchmarks.get(scenario, {})
        old_data = old_benchmarks.get(scenario, {})
        
        print(f"Scenario: {scenario}")
        
        if 'rps' in new_data and 'rps' in old_data:
            change = ((new_data['rps'] - old_data['rps']) / old_data['rps']) * 100
            print(f"  RPS: {old_data['rps']:.2f} → {new_data['rps']:.2f} ({change:+.2f}%)")
        
        if 'latency_avg' in new_data and 'latency_avg' in old_data:
            change = ((new_data['latency_avg'] - old_data['latency_avg']) / old_data['latency_avg']) * 100
            print(f"  Latency: {old_data['latency_avg']:.2f}ms → {new_data['latency_avg']:.2f}ms ({change:+.2f}%)")
        
        print()


def main():
    parser = argparse.ArgumentParser(description='Update performance baseline')
    parser.add_argument('--results', '-r', required=True, help='Path to performance results JSON')
    parser.add_argument('--baseline', '-b', default='docs/performance/baseline.json', help='Path to baseline file')
    parser.add_argument('--history', '-H', default='docs/performance/baseline-history.json', help='Path to history file')
    parser.add_argument('--version', '-v', help='Version string (default: auto-detect)')
    parser.add_argument('--commit', '-c', help='Commit SHA (default: auto-detect)')
    parser.add_argument('--no-backup', action='store_true', help='Skip backup of old baseline')
    parser.add_argument('--no-history', action='store_true', help='Skip updating history file')
    parser.add_argument('--max-history', type=int, default=30, help='Maximum history entries')
    parser.add_argument('--compare', action='store_true', help='Compare with old baseline')
    
    args = parser.parse_args()
    
    # 加载结果
    results = load_current_results(args.results)
    if not results:
        print("Failed to load results", file=sys.stderr)
        sys.exit(1)
    
    # 提取版本和提交信息
    version = args.version
    commit = args.commit
    
    if not version:
        # 尝试从结果中获取版本
        version = results.get('version', 'unknown')
    
    if not commit:
        # 尝试从结果中获取提交信息
        if 'commit' in results:
            commit = results['commit'].get('sha', 'unknown')
        else:
            commit = 'unknown'
    
    # 提取基线数据
    new_baseline = extract_baseline_data(results, version, commit)
    if not new_baseline:
        print("Failed to extract baseline data", file=sys.stderr)
        sys.exit(1)
    
    # 加载旧基线
    old_baseline = None
    if args.compare or os.path.exists(args.baseline):
        old_baseline = load_baseline(args.baseline)
    
    # 比较基线
    if args.compare and old_baseline:
        compare_with_baseline(new_baseline, old_baseline)
    
    # 更新基线文件
    baseline_dir = os.path.dirname(args.baseline)
    if baseline_dir:
        os.makedirs(baseline_dir, exist_ok=True)
    
    update_baseline_file(args.baseline, new_baseline, backup=not args.no_backup)
    
    # 更新历史文件
    if not args.no_history:
        history_dir = os.path.dirname(args.history)
        if history_dir:
            os.makedirs(history_dir, exist_ok=True)
        
        update_history_file(args.history, new_baseline, max_entries=args.max_history)
    
    print("\n✅ Baseline update completed successfully")


if __name__ == '__main__':
    main()

File: scripts/update_baseline.js
#!/usr/bin/env node
/**
 * Update performance baseline script
 * 
 * This script updates the performance baseline file with current test results
 */

const fs = require('fs').promises;
const path = require('path');

/**
 * Load current performance test results
 * @param {string} resultsFile - Path to results file
 * @returns {Promise<Object|null>} Parsed results or null
 */
async function loadCurrentResults(resultsFile) {
  try {
    const content = await fs.readFile(resultsFile, 'utf-8');
    return JSON.parse(content);
  } catch (error) {
    console.error(`Error loading results: ${error.message}`);
    return null;
  }
}

/**
 * Load current baseline
 * @param {string} baselineFile - Path to baseline file
 * @returns {Promise<Object|null>} Parsed baseline or null
 */
async function loadBaseline(baselineFile) {
  try {
    const content = await fs.readFile(baselineFile, 'utf-8');
    return JSON.parse(content);
  } catch (error) {
    console.error(`Baseline file not found, will create new: ${error.message}`);
    return null;
  }
}

/**
 * Extract baseline data from results
 * @param {Object} results - Performance test results
 * @param {string} version - Version string
 * @param {string} commit - Commit SHA
 * @returns {Object|null} Baseline data or null
 */
function extractBaselineData(results, version, commit) {
  if (!results) {
    return null;
  }
  
  const baselineData = {
    version: version || 'unknown',
    date: new Date().toISOString(),
    commit: commit || 'unknown',
    baseline: {}
  };
  
  // Extract RPS data
  if (results.test_scenarios) {
    for (const scenario of results.test_scenarios) {
      const scenarioName = scenario.name || 'unknown';
      
      if (scenario.results) {
        const scenarioData = {};
        
        if (scenario.results.rps) {
          scenarioData.rps = scenario.results.rps.value;
        }
        
        if (scenario.results.latency_avg) {
          const latency = scenario.results.latency_avg.value;
          if (latency !== 'N/A') {
            scenarioData.latency_avg = latency;
          }
        }
        
        if (Object.keys(scenarioData).length > 0) {
          baselineData.baseline[scenarioName] = scenarioData;
        }
      }
    }
  }
  
  // Extract environment information
  if (results.environment) {
    baselineData.environment = {
      os: results.environment.os || 'unknown',
      runner: results.environment.runner || 'unknown',
      cpu: results.environment.cpu || 'unknown',
      compiler: results.environment.compiler || 'unknown',
      compiler_version: results.environment.compiler_version || 'unknown',
      build_type: results.environment.build_type || 'Release',
      optimization: results.environment.optimization || '-O2'
    };
  }
  
  return baselineData;
}

/**
 * Update baseline file
 * @param {string} baselineFile - Path to baseline file
 * @param {Object} newBaseline - New baseline data
 * @param {boolean} backup - Whether to backup old baseline
 * @returns {Promise<boolean>} Success status
 */
async function updateBaselineFile(baselineFile, newBaseline, backup = true) {
  try {
    // Backup old baseline
    if (backup) {
      try {
        await fs.copyFile(baselineFile, `${baselineFile}.backup`);
        console.log(`Baseline backed up to ${baselineFile}.backup`);
      } catch (error) {
        // File doesn't exist, skip backup
      }
    }
    
    // Write new baseline
    await fs.writeFile(baselineFile, JSON.stringify(newBaseline, null, 2), 'utf-8');
    
    console.log(`✅ Baseline updated: ${baselineFile}`);
    console.log(`   Version: ${newBaseline.version}`);
    console.log(`   Date: ${newBaseline.date}`);
    
    return true;
  } catch (error) {
    console.error(`Error updating baseline: ${error.message}`);
    return false;
  }
}

/**
 * Update history file
 * @param {string} historyFile - Path to history file
 * @param {Object} newBaseline - New baseline data
 * @param {number} maxEntries - Maximum history entries
 * @returns {Promise<boolean>} Success status
 */
async function updateHistoryFile(historyFile, newBaseline, maxEntries = 30) {
  try {
    let history = [];
    
    // Load existing history
    try {
      const content = await fs.readFile(historyFile, 'utf-8');
      history = JSON.parse(content);
    } catch (error) {
      history = [];
    }
    
    // Add new entry
    history.unshift(newBaseline);
    
    // Limit entries
    history = history.slice(0, maxEntries);
    
    // Write history file
    await fs.writeFile(historyFile, JSON.stringify(history, null, 2), 'utf-8');
    
    console.log(`✅ History updated: ${historyFile} (${history.length} entries)`);
    
    return true;
  } catch (error) {
    console.error(`Error updating history: ${error.message}`);
    return false;
  }
}

/**
 * Compare new baseline with old baseline
 * @param {Object} newBaseline - New baseline data
 * @param {Object} oldBaseline - Old baseline data
 */
function compareWithBaseline(newBaseline, oldBaseline) {
  if (!oldBaseline) {
    console.log('No old baseline found for comparison');
    return;
  }
  
  console.log('\n=== Baseline Comparison ===\n');
  
  const newBenchmarks = newBaseline.baseline || {};
  const oldBenchmarks = oldBaseline.baseline || {};
  
  const allScenarios = new Set([
    ...Object.keys(newBenchmarks),
    ...Object.keys(oldBenchmarks)
  ]);
  
  for (const scenario of Array.from(allScenarios).sort()) {
    const newData = newBenchmarks[scenario] || {};
    const oldData = oldBenchmarks[scenario] || {};
    
    console.log(`Scenario: ${scenario}`);
    
    if ('rps' in newData && 'rps' in oldData) {
      const change = ((newData.rps - oldData.rps) / oldData.rps) * 100;
      console.log(`  RPS: ${oldData.rps.toFixed(2)} → ${newData.rps.toFixed(2)} (${change >= 0 ? '+' : ''}${change.toFixed(2)}%)`);
    }
    
    if ('latency_avg' in newData && 'latency_avg' in oldData) {
      const change = ((newData.latency_avg - oldData.latency_avg) / oldData.latency_avg) * 100;
      console.log(`  Latency: ${oldData.latency_avg.toFixed(2)}ms → ${newData.latency_avg.toFixed(2)}ms (${change >= 0 ? '+' : ''}${change.toFixed(2)}%)`);
    }
    
    console.log('');
  }
}

/**
 * Main function
 */
async function main() {
  const args = process.argv.slice(2);
  
  // Parse arguments
  const options = {
    results: null,
    baseline: 'docs/performance/baseline.json',
    history: 'docs/performance/baseline-history.json',
    version: null,
    commit: null,
    noBackup: false,
    noHistory: false,
    maxHistory: 30,
    compare: false
  };
  
  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case '--results':
      case '-r':
        options.results = args[++i];
        break;
      case '--baseline':
      case '-b':
        options.baseline = args[++i];
        break;
      case '--history':
      case '-H':
        options.history = args[++i];
        break;
      case '--version':
      case '-v':
        options.version = args[++i];
        break;
      case '--commit':
      case '-c':
        options.commit = args[++i];
        break;
      case '--no-backup':
        options.noBackup = true;
        break;
      case '--no-history':
        options.noHistory = true;
        break;
      case '--max-history':
        options.maxHistory = parseInt(args[++i], 10);
        break;
      case '--compare':
        options.compare = true;
        break;
      default:
        console.error(`Unknown option: ${args[i]}`);
        process.exit(1);
    }
  }
  
  if (!options.results) {
    console.error('Error: --results is required');
    console.error('');
    console.error('Usage: node update_baseline.js --results <performance_results.json> [options]');
    console.error('');
    console.error('Options:');
    console.error('  --results, -r    Path to performance results JSON (required)');
    console.error('  --baseline, -b   Path to baseline file (default: docs/performance/baseline.json)');
    console.error('  --history, -H    Path to history file (default: docs/performance/baseline-history.json)');
    console.error('  --version, -v     Version string (default: auto-detect)');
    console.error('  --commit, -c      Commit SHA (default: auto-detect)');
    console.error('  --no-backup       Skip backup of old baseline');
    console.error('  --no-history      Skip updating history file');
    console.error('  --max-history N   Maximum history entries (default: 30)');
    console.error('  --compare         Compare with old baseline');
    console.error('');
    console.error('Example:');
    console.error('  node update_baseline.js --results performance.json');
    process.exit(1);
  }
  
  // Load results
  const results = await loadCurrentResults(options.results);
  if (!results) {
    console.error('Failed to load results');
    process.exit(1);
  }
  
  // Extract version and commit info
  let version = options.version;
  let commit = options.commit;
  
  if (!version) {
    version = results.version || 'unknown';
  }
  
  if (!commit) {
    if (results.commit) {
      commit = results.commit.sha || 'unknown';
    } else {
      commit = 'unknown';
    }
  }
  
  // Extract baseline data
  const newBaseline = extractBaselineData(results, version, commit);
  if (!newBaseline) {
    console.error('Failed to extract baseline data');
    process.exit(1);
  }
  
  // Load old baseline
  const oldBaseline = options.compare || await loadBaseline(options.baseline);
  
  // Compare baseline
  if (options.compare && oldBaseline) {
    compareWithBaseline(newBaseline, oldBaseline);
  }
  
  // Update baseline file
  const baselineDir = path.dirname(options.baseline);
  if (baselineDir) {
    await fs.mkdir(baselineDir, { recursive: true });
  }
  
  await updateBaselineFile(options.baseline, newBaseline, !options.noBackup);
  
  // Update history file
  if (!options.noHistory) {
    const historyDir = path.dirname(options.history);
    if (historyDir) {
      await fs.mkdir(historyDir, { recursive: true });
    }
    
    await updateHistoryFile(options.history, newBaseline, options.maxHistory);
  }
  
  console.log('\n✅ Baseline update completed successfully');
}

main().catch(error => {
  console.error('Error:', error.message);
  process.exit(1);
});